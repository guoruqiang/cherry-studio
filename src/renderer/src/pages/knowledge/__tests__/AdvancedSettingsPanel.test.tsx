import type { KnowledgeBase, Model } from '@renderer/types'
import { render, screen } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event'
import { beforeEach, describe, expect, it, vi } from 'vitest'

import AdvancedSettingsPanel from '../components/KnowledgeSettings/AdvancedSettingsPanel'

const mocks = vi.hoisted(() => {
  return {
    i18n: {
      t: (k: string) => {
        const translations: Record<string, string> = {
          'knowledge.chunk_size': '分块大小',
          'knowledge.chunk_overlap': '分块重叠',
          'knowledge.threshold': '检索相似度阈值',
          'knowledge.chunk_size_change_warning': '避免修改这个高级设置。'
        }
        return translations[k] || k
      }
    },
    handlers: {
      handleChunkSizeChange: vi.fn(),
      handleChunkOverlapChange: vi.fn(),
      handleThresholdChange: vi.fn()
    }
  }
})

vi.mock('@renderer/components/InfoTooltip', () => ({
  default: ({ title }: { title: string }) => <div>{mocks.i18n.t(title)}</div>
}))

vi.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: mocks.i18n.t
  })
}))

/**
 * 创建测试用的 KnowledgeBase 对象
 * @param overrides 可选的属性覆盖
 * @returns KnowledgeBase 对象
 */
function createKnowledgeBase(overrides: Partial<KnowledgeBase> = {}): KnowledgeBase {
  return {
    id: '1',
    name: 'Test KB',
    model: {
      id: 'test-model',
      provider: 'test-provider',
      name: 'Test Model',
      group: 'test'
    } as Model,
    items: [],
    created_at: Date.now(),
    updated_at: Date.now(),
    version: 1,
    chunkSize: 500,
    chunkOverlap: 200,
    threshold: 0.5,
    ...overrides
  }
}

describe('AdvancedSettingsPanel', () => {
  const mockBase = createKnowledgeBase()

  beforeEach(() => {
    vi.clearAllMocks()
  })

  describe('basic rendering', () => {
    it('should match snapshot', () => {
      const { container } = render(<AdvancedSettingsPanel newBase={mockBase} handlers={mocks.handlers} />)

      expect(container.firstChild).toMatchSnapshot()
    })
  })

  describe('handlers', () => {
    it('should call handlers when values are changed', async () => {
      const user = userEvent.setup()
      render(<AdvancedSettingsPanel newBase={mockBase} handlers={mocks.handlers} />)

      const chunkSizeInput = screen.getByLabelText('分块大小')
      await user.clear(chunkSizeInput)
      await user.type(chunkSizeInput, '600')
      expect(mocks.handlers.handleChunkSizeChange).toHaveBeenCalledWith(600)

      const chunkOverlapInput = screen.getByLabelText('分块重叠')
      await user.clear(chunkOverlapInput)
      await user.type(chunkOverlapInput, '300')
      expect(mocks.handlers.handleChunkOverlapChange).toHaveBeenCalledWith(300)

      const thresholdInput = screen.getByLabelText('检索相似度阈值')
      await user.clear(thresholdInput)
      await user.type(thresholdInput, '0.6')
      expect(mocks.handlers.handleThresholdChange).toHaveBeenCalledWith(0.6)
    })
  })
})
