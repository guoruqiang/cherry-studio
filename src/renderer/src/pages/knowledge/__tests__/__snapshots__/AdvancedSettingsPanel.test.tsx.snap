// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AdvancedSettingsPanel > basic rendering > should match snapshot 1`] = `
.c0 {
  padding: 0 16px;
}

.c1 {
  margin-bottom: 24px;
}

.c1 .settings-label {
  font-size: 14px;
  margin-bottom: 8px;
  display: flex;
  align-items: center;
  gap: 8px;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="settings-label"
    >
      分块大小
      <div>
        knowledge.chunk_size_tooltip
      </div>
    </div>
    <div
      class="ant-input-number css-dev-only-do-not-override-1261szd ant-input-number-outlined"
      style="width: 100%;"
    >
      <div
        class="ant-input-number-handler-wrap"
      >
        <span
          aria-disabled="false"
          aria-label="Increase Value"
          class="ant-input-number-handler ant-input-number-handler-up"
          role="button"
          unselectable="on"
        >
          <span
            aria-label="up"
            class="anticon anticon-up ant-input-number-handler-up-inner"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="up"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
              />
            </svg>
          </span>
        </span>
        <span
          aria-disabled="false"
          aria-label="Decrease Value"
          class="ant-input-number-handler ant-input-number-handler-down"
          role="button"
          unselectable="on"
        >
          <span
            aria-label="down"
            class="anticon anticon-down ant-input-number-handler-down-inner"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="down"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
              />
            </svg>
          </span>
        </span>
      </div>
      <div
        class="ant-input-number-input-wrap"
      >
        <input
          aria-label="分块大小"
          aria-valuemin="100"
          aria-valuenow="500"
          autocomplete="off"
          class="ant-input-number-input"
          placeholder="knowledge.chunk_size_placeholder"
          role="spinbutton"
          step="1"
          value="500"
        />
      </div>
    </div>
  </div>
  <div
    class="c1"
  >
    <div
      class="settings-label"
    >
      分块重叠
      <div>
        knowledge.chunk_overlap_tooltip
      </div>
    </div>
    <div
      class="ant-input-number css-dev-only-do-not-override-1261szd ant-input-number-outlined"
      style="width: 100%;"
    >
      <div
        class="ant-input-number-handler-wrap"
      >
        <span
          aria-disabled="false"
          aria-label="Increase Value"
          class="ant-input-number-handler ant-input-number-handler-up"
          role="button"
          unselectable="on"
        >
          <span
            aria-label="up"
            class="anticon anticon-up ant-input-number-handler-up-inner"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="up"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
              />
            </svg>
          </span>
        </span>
        <span
          aria-disabled="false"
          aria-label="Decrease Value"
          class="ant-input-number-handler ant-input-number-handler-down"
          role="button"
          unselectable="on"
        >
          <span
            aria-label="down"
            class="anticon anticon-down ant-input-number-handler-down-inner"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="down"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
              />
            </svg>
          </span>
        </span>
      </div>
      <div
        class="ant-input-number-input-wrap"
      >
        <input
          aria-label="分块重叠"
          aria-valuemin="0"
          aria-valuenow="200"
          autocomplete="off"
          class="ant-input-number-input"
          placeholder="knowledge.chunk_overlap_placeholder"
          role="spinbutton"
          step="1"
          value="200"
        />
      </div>
    </div>
  </div>
  <div
    class="c1"
  >
    <div
      class="settings-label"
    >
      检索相似度阈值
      <div>
        knowledge.threshold_tooltip
      </div>
    </div>
    <div
      class="ant-input-number css-dev-only-do-not-override-1261szd ant-input-number-outlined"
      style="width: 100%;"
    >
      <div
        class="ant-input-number-handler-wrap"
      >
        <span
          aria-disabled="false"
          aria-label="Increase Value"
          class="ant-input-number-handler ant-input-number-handler-up"
          role="button"
          unselectable="on"
        >
          <span
            aria-label="up"
            class="anticon anticon-up ant-input-number-handler-up-inner"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="up"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
              />
            </svg>
          </span>
        </span>
        <span
          aria-disabled="false"
          aria-label="Decrease Value"
          class="ant-input-number-handler ant-input-number-handler-down"
          role="button"
          unselectable="on"
        >
          <span
            aria-label="down"
            class="anticon anticon-down ant-input-number-handler-down-inner"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="down"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
              />
            </svg>
          </span>
        </span>
      </div>
      <div
        class="ant-input-number-input-wrap"
      >
        <input
          aria-label="检索相似度阈值"
          aria-valuemax="1"
          aria-valuemin="0"
          aria-valuenow="0.5"
          autocomplete="off"
          class="ant-input-number-input"
          placeholder="knowledge.threshold_placeholder"
          role="spinbutton"
          step="0.1"
          value="0.5"
        />
      </div>
    </div>
  </div>
  <div
    class="ant-alert ant-alert-warning css-dev-only-do-not-override-1261szd"
    data-show="true"
    role="alert"
  >
    <span
      aria-label="warning"
      class="anticon anticon-warning ant-alert-icon"
      role="img"
    >
      <svg
        aria-hidden="true"
        data-icon="warning"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M464 720a48 48 0 1096 0 48 48 0 10-96 0zm16-304v184c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V416c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8zm475.7 440l-416-720c-6.2-10.7-16.9-16-27.7-16s-21.6 5.3-27.7 16l-416 720C56 877.4 71.4 904 96 904h832c24.6 0 40-26.6 27.7-48zm-783.5-27.9L512 239.9l339.8 588.2H172.2z"
        />
      </svg>
    </span>
    <div
      class="ant-alert-content"
    >
      <div
        class="ant-alert-message"
      >
        避免修改这个高级设置。
      </div>
    </div>
  </div>
</div>
`;
